#PBS -N pan_repeatmask
#PBS -l nodes=node22:ppn=20,walltime=720:00:00:00,mem=100g
#PBS -q defaultApp
#PBS -j oe
#PBS -o log_repeat_maskerV5

echo "Starting"
date
hostname

cd $PBS_O_WORKDIR

#export LD_LIBRARY_PATH=/usr/local/lib:/usr/lib64:$LD_LIBRARY_PATH
##set arguments
#genome="/public4/home/huangyj/tobacco/assembly/05gap_close/seprate_genome_v5/HDR7_genome_version5_full.fasta"
filename=${genome##*/}
speciesname=${filename%%.*}
ltr="/public/home/huangyj/anaconda3/envs/LAI/bin/LTR_retriever"

source activate /public/home/huangyj/anaconda3/envs/repeatscout
##job area
##Step1
rm -rf ./RM*
BuildDatabase -name $speciesname -engine ncbi $genome
RepeatModeler -threads 10 -database $speciesname -LTRStruct

#cd RM_150467.TueOct31648582023
#/public/home/huangyj/software/RepeatModeler-master/RepeatClassifier -consensi consensi.fa -stockholm families.stk
#cd ../

#LTRPipeline -threads 50 -ltr_retriever_dir $ltr $genome

echo "base build"
##Step2
#lib_dir=`grep "Using output directory" $log | cut -f 5 -d " "`
lib_dir=`ls |grep -E "^RM_"`
lib="$lib_dir/consensi.fa.classified"
#lib="HDR7_genome_version5_full-families.fa"
##argument -lib: produced by steps above.
if [ -f $lib ] ; then
:
#RepeatMasker -pa 8 -qq -e ncbi -lib $lib -gff $genome -dir ./
fi
conda deactivate
echo "Finished"
date

